File: blueprints/equipment/equipment.py
from extensions import (
    Blueprint, request, jsonify, abort, db, login_required, role_required,
    current_user_id,
)
from models import Equipment, EquipmentRequest
from datetime import datetime

equipmentBp = Blueprint("equipment", __name__, url_prefix="/api/equipment")

# Specialised equipment requests come from staff (and admins acting as staff).
REQUESTERS = ("staff", "admin")
MODIFIERS = ("manager",)

@equipmentBp.get("")
@login_required
def list_equipment():
    """Catalogue of equipment (drop-down options for the request form)."""
    return jsonify(equipment=[e.to_dict() for e in Equipment.query.all()])


@equipmentBp.post("/request")
@role_required(*REQUESTERS)
def submit_request():
    """Submit a specialised equipment request (Equipment Request form)."""
    data = request.get_json(silent=True) or {}
    name = (data.get("equipmentName") or "").strip()
    if not name:
        abort(400, description="'equipmentName' is required.")

    requested_date = None
    if data.get("requestedDate"):
        try:
            requested_date = datetime.fromisoformat(data["requestedDate"])
        except ValueError:
            abort(400, description="Invalid 'requestedDate'.")

    req = EquipmentRequest(
        userId=current_user_id(),
        equipmentName=name,
        purpose=data.get("purpose"),
        requestedDate=requested_date,
        attendees=data.get("attendees"),
        status="pending",
    )
    db.session.add(req)

    # Notify the requester that the approval workflow has started.
    from blueprints.booking.services import notify
    notify(current_user_id(), "Request submitted",
           f"Your request for {name} is pending manager approval.",
           type="approval")
    db.session.commit()
    return jsonify(request=req.to_dict()), 201


@equipmentBp.get("/request")
@login_required
def my_requests():
    """List the current user's equipment requests."""
    rows = (
        EquipmentRequest.query.filter_by(userId=current_user_id())
        .order_by(EquipmentRequest.created_at.desc())
        .all()
    )
    return jsonify(requests=[r.to_dict() for r in rows])

File: static/js/equipment.js
/* Specialized Equipment: populate the catalogue and submit a request. */
(function () {
  async function loadCatalogue() {
    try {
      var data = await api.get("/api/equipment");
      var sel = byId("eq_name");
      if (!data.equipment.length) { sel.innerHTML = '<option>No equipment available</option>'; return; }
      sel.innerHTML = data.equipment.map(function (e) {
        return '<option>' + e.name + '</option>';
      }).join("");
    } catch (e) {}
  }

  byId("equipmentForm").addEventListener("submit", async function (e) {
    e.preventDefault();
    var body = {
      equipmentName: byId("eq_name").value,
      purpose: byId("eq_purpose").value,
      attendees: parseInt(byId("eq_attendees").value) || null
    };
    if (byId("eq_date").value) body.requestedDate = byId("eq_date").value + "T09:00";
    try {
      await api.post("/api/equipment/request", body);
      showToast("Request submitted", "Your request is pending manager approval.");
    } catch (err) { showToast("Could not submit", err.message); }
  });

  document.addEventListener("DOMContentLoaded", loadCatalogue);
})();
